src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,585 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	char *buf_start,
+	char **buf_next);
+
+static
+size_t
+write_event_buffer_int32 (
+	int32_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+size_t
+write_event_buffer_int64 (
+	int64_t val,
+	char *buf_start,
+	char **buf_next);
+
+static
+uint64_t
+get_typeid_for_type (MonoType *t);
+
+static
+uint64_t
+get_typeid_for_class (MonoClass *c);
+
+// Clear out BulkTypeValue before filling it out (array elements can get reused if there
+// are enough types that we need to flush to multiple events).
+static
+void
+ep_rt_bulk_type_value_clear (BulkTypeValue *bulk_type_value)
+{
+	memset (bulk_type_value, 0, sizeof(BulkTypeValue));
+}
+
+static
+int
+ep_rt_mono_get_byte_count_in_event (BulkTypeValue *bulk_type_value)
+{
+	int name_len = 0;
+
+	return sizeof (bulk_type_value->fixed_sized_data.type_id) + 	// Fixed Sized Data
+		sizeof (bulk_type_value->fixed_sized_data.module_id) +
+		sizeof (bulk_type_value->fixed_sized_data.type_name_id) +
+		sizeof (bulk_type_value->fixed_sized_data.flags) +
+		sizeof (bulk_type_value->fixed_sized_data.cor_element_type) +
+		sizeof (bulk_type_value->type_parameters_count) +		// Type parameters
+		(name_len + 1) * sizeof (ep_char8_t) +		// Size of name, including null terminator
+		bulk_type_value->type_parameters_count * sizeof (uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+#define MAX_EVENT_BYTE_COUNT (63 * 1024)
+
+// The maximum event size, and the size of the buffer that we allocate to hold the event contents.
+#define MAX_SIZE_OF_EVENT_BUFFER 65536
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array. (Intentionally overestimate the size of the non-array parts to keep it safe.)
+// This follows CoreCLR's kMaxBytesTypeValues.
+#define MAX_TYPE_VALUES_BYTES (MAX_EVENT_BYTE_COUNT - 0x30)
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+#define K_MAX_COUNT_TYPE_VALUES ((uint32_t)(128 - 2) / 4)
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue bulk_type_values [K_MAX_COUNT_TYPE_VALUES];
+	uint8_t *bulk_type_event_buffer;
+	uint32_t bulk_type_value_count;
+	uint32_t bulk_type_value_byte_count;
+};
```

Comment:
I would just add the mempool as an extra member here since its lifetime is tied to the BulkTypeEventLogger.  Then you don't have to pass the mempool as an extra argument to `ep_rt_mono_log_type_and_parameters` and other functions.

Reply:
Moved the dedicated memory pool to be a BulkTypeEventLogger field! Thanks